Add Redox support for most of the modules

AdminXVII opened the pull request:
Some things are not implemented yet in redox, so a lot of annotations were added to remove functions when compiling for redox. Those functions will hopefully be added in time, but for now it's better to have partial support than none.

Blocked by https://github.com/rust-lang/libc/pull/1438

asomers commented:
Interesting!  But I have two big questions:
1) What's the best way to get Redox running in a VM?  The last time I tried I couldn't get it to work.
2) How would you test on Redox in CI?

AdminXVII commented:
There is [redoxer](https://gitlab.redox-os.org/redox-os/redoxer) that does exactly this. You basically pull the redoxos/redoxer docker image and then run `redoxer test`

AdminXVII commented:
https://gitlab.redox-os.org/redox-os/redoxfs/blob/master/.gitlab-ci.yml is an example

AdminXVII commented:
The libc request was merged.

mati865 commented on Cargo.toml:
```diff
@@ -16,7 +16,7 @@ exclude     = [
 
 [dependencies]
-libc = { version = "0.2.60", features = [ "extra_traits" ] }
+libc = { git = "https://github.com/rust-lang/libc", features = [ "extra_traits" ] } # TODO: switch to 0.2.63 when it's out
```
`0.2.63` was yanked.

AdminXVII replied:
Right, thanks. Fixed the comment.

AdminXVII commented:
Btw, should I include the tests for redox in this PR or should I make another one specifically for it?

asomers commented:
Best if you can add them to this PR.

AdminXVII commented:
@asomers the CI succeeded and I only added a new test. Is the problem resolved?

asomers commented:
Amazing!  You fixed it! ;).  I guess Travis undid whatever they did on September 6.  I see a lot of warnings in the Redox test build, though.  Could you please silence all of the warnings?

AdminXVII commented:
I was using the warnings to remember what needs to be included. I'd have silenced them when you would have been ready to merge. Is that ok with you?

asomers commented:
You mean you want to silence them after I've reviewed the rest of the PR?  That's ok.

AdminXVII commented:
There is currently ongoing work to improve what redox can do, so as this goes I remove `cfg`s to add more things for Redox. I used the comment to see what was needed to add. That said I think it's better if I remove and simply `grep`, so discard my previous comment.

asomers commented:
Rebasing should fix the test failures now.

AdminXVII commented:
Just force-pushed.

AdminXVII commented:
@asomers Pinging to see if everything's fine. I just rebased.

asomers commented:
Yeah, I just need to find time to review it.  I'll try to do it during my Christmas vacation.  And one more question: redoxer is good for CI tests, but what's the best way to create a persistent Redox VM for development?

AdminXVII commented:
You should clone the redox repo and follow the instructions. This will setup a QEMU environment. For development though, redoxer is really useful, as you can just `redoxer run` and it will run the crate inside redox with low effort.

asomers commented:
I can't get Redox to build.  There are Linuxisms in the build process.  Are there prebuilt binary images somewhere?  I can't find any.

AdminXVII commented:
Not to my knowledge, sorry

cdbattags commented:
Any insight into why 1.37 is failing?

asomers commented:
> Any insight into why 1.37 is failing?

That test has been failing on Linux intermittently for awhile.  I suspect a glibc bug, but I've never been able to prove it.

AdminXVII commented:
@asomers Is there something you feel is missing on this PR for merge? I just wanted to make sure you knew I personally think it's ready

asomers commented:
Just that I haven't been able to test it yet, because there are no pre-built images and Redox doesn't build on FreeBSD.  Does Redox really not publish pre-built images anywhere?

AdminXVII commented:
It's work in progress. There is an ongoing discussion to create container images, but it's not yet finished. I think you could exec a shell inside a local `redoxos/redoxer` container in docker to test, though.

asomers commented:
Link to progress of building Redox on FreeBSD: https://gitlab.redox-os.org/redox-os/redox/issues/952

AdminXVII commented:
@asomers I've rebased my changes. I've also fixed compilation of redox on FreeBSD. You will need to disable compilation of netsurf (comment it in `config/desktop.toml`), but after that master should compile and run at least with `gmake qemu vga=none`.

If you don't have the time to review it, would it be ok with you to merge this nonetheless? A lot of the userspace programs depend on this to start compiling for Redox, so the faster this gets merged, the better it will be for us. As for testing before merging, I totally understand that you want to uphold the same standard of quality, but I think it would be fine to say that if something breaks, it's our fault, not yours and that the Redox team will deal with it. It'd be a lot better for us to fix the occasional bug than to have half of the ecosystem that simply can't compile.

Thanks for your great work and cheers!

asomers commented:
I saw that comment over on Gitlab.  Sounds awesome!  But I have two questions:
1) Is QEMU necessary for compiling Redox, or only for running it?
2) Has anybody ever tried running Redox in BHyve?

AdminXVII commented:
1- Qemu is only necessary to run it. To compile, you'd simply do `gmake all`.
2- Not that I'm aware of.

asomers commented on src/fcntl.rs:
```diff
@@ -153,21 +161,21 @@ libc_bitflags!(
     let fd = path.with_nix_path(|cstr| {
-        let modebits = c_uint::from(mode.bits());
-        unsafe { libc::open(cstr.as_ptr(), oflag.bits(), modebits) }
+        unsafe { libc::open(cstr.as_ptr(), oflag.bits(), mode.bits() as c_uint) }
```
Why did you make this change?  Using `from` is usually preferable to `as`.

AdminXVII replied:
It's because mode_t is a signed int on redox, so in terms of integer, because it can be lossy, the `From<c_uint>` trait is not implemented for `c_int` but `as c_uint` works fine.

asomers commented on src/fcntl.rs:
```diff
@@ -153,21 +161,21 @@ libc_bitflags!(
     let fd = path.with_nix_path(|cstr| {
-        let modebits = c_uint::from(mode.bits());
-        unsafe { libc::openat(dirfd, cstr.as_ptr(), oflag.bits(), modebits) }
+        unsafe { libc::openat(dirfd, cstr.as_ptr(), oflag.bits(), mode.bits() as c_uint) }
```
Ditto from/as.

asomers commented on test/sys/test_signal.rs:
```diff
@@ -99,9 +109,6 @@ fn test_signal() {
     assert!(SIGNALED.load(Ordering::Relaxed));
     assert_eq!(unsafe { signal(Signal::SIGINT, SigHandler::SigDfl) }.unwrap(), handler);
 
-    let action_handler = SigHandler::SigAction(test_sigaction_action);
```
Ok, this was dumb on our part to have all of that different functionality in a single test.  Thanks for splitting it up a bit.

asomers commented on src/dir.rs:
```diff
@@ -1,15 +1,25 @@
+#[cfg(not(target_os = "redox"))]
```
It looks like the only thing in this file that's still defined on Redox is `Entry`.  But what good is `Entry` without the other parts?  I think you should revert all changes to this file, and simply cfg out the entire module from src/lib.rs.

AdminXVII replied:
You are right, changed it.

AdminXVII commented:
The `as` cast is needed because mode_t is signed on redox while unsigned on open. From<c_uint> is not implemented for c_int, but it's not a problem since this will leave the bits in place, which is the expected behaviour.

AdminXVII commented:
Rebased

asomers commented:
Ok, could you please do one last thing?  Just add Redox to the list of supported platforms in the README.md.  Put it in the "Tier 3" section.

AdminXVII commented:
Sure it's done!

AdminXVII commented:
Yes! Thanks a lot for your time, it's greatly appreciated!
